File: knightsMove.py
class Board:
	def __init__(self):
		self.board = []
		for i in range(8):
			col = []
			for j in range(8):
				col.append(0)
			self.board.append(col)

	def minSteps(self, sourceX, sourceY, destX, destY):

		if(sourceX >= 8 or sourceX < 0 or sourceY >= 8 or sourceX < 0):
			return -1

		if(sourceX == destX and sourceY == destY):
			return 0
		
		step = 0

		dx = [1,1,-1,-1,2,2,-2,-2]
		dy = [2,-2,2,-2,1,-1,1,-1]

		queue = []
		addStep = True

		queue.append((sourceX,sourceY))
		self.board[sourceX][sourceY] = 1
		queue.append(addStep)

		while queue:
			front = queue.pop(0)
			if type(front) == bool:
				if(queue):
					queue.append(addStep)
					step = step + 1
				else:
					break
			else:
				currX = front[0]
				currY = front[1]

				if currX == destX and currY == destY:
					return step

				for i in range(8):
					nextX = currX + dx[i]
					nextY = currY + dy[i]
					if(nextX >= 0 and nextX < 8 and nextY >=0 and nextY < 8 and self.board[nextX][nextY]==0):
						queue.append((nextX,nextY))
						self.board[nextX][nextY] = 1


chess = Board();

sourceX = int(input("\nEnter starting X position of the Knight: "))
sourceY = int(input("\nEnter starting Y position of the Knight: "))
destX = int(input("\nEnter final X position of the Knight: "))
destY = int(input("\nEnter final Y position of the Knight: "))

print(chess.minSteps(sourceX,sourceY,destX,destY))

File: knightsMove.cpp
#include<iostream>
#include<queue>
#include<vector>
using namespace std;

class Board {
	int n;
	vector<vector<int> > board;
public:
	Board(int size) {
		n = size;
		vector<int> temp(n, 0);
		for (int i = 0; i < n; i++) {
			board.push_back(temp);
		}
	}

	int minSteps(int sourceX, int sourceY, int destX, int destY) {
		if (sourceX == destX && sourceY == destY) {
			return 0;
		}

		int step = 0;

		int dx[] = {1, 1, -1, -1, 2, 2, -2, -2};
		int dy[] = {2, -2, 2, -2, 1, -1, 1, -1};

		queue<pair<int, int>> q;
		q.push({sourceX, sourceY});
		q.push({ -1, -1});
		board[sourceX][sourceY] = 1;

		while (!q.empty()) {
			pair<int, int> pr = q.front();
			q.pop();
			if (pr.first == -1 && pr.second == -1) {
				step++;
				if (q.empty()) {
					break;
				}
				q.push({ -1, -1});
				continue;
			}

			int x = pr.first;
			int y = pr.second;

			if (x == destX && y == destY) {
				return step;
			}

			for (int i = 0; i < 8; i++) {
				if (x + dx[i] >= 0 && x + dx[i] < n && y + dy[i] >= 0 && y + dy[i] < n && board[x + dx[i]][y + dy[i]] == 0) {
					q.push({x + dx[i], y + dy[i]});
					board[x + dx[i]][y + dy[i]] = 1;
				}
			}

		}

		return -1;


	}

};

int main() {
	Board chess(8);
	cout << chess.minSteps(7, 0, 7, 5);
	return 0;
}
